### DISCUSSION TITLE
How should I best define scanner.c if my grammar inherits from a base grammar with external rules?

### DISCUSSION THREAD

**Discussion by michaelfortunato** [tree-sitter/tree-sitter#3142] [Q&A]
Say I have a grammar which has a few external rules defined.
Say I then the base grammar for a new grammar by doing 
```javascript
module.exports = grammar(baseGrammar, { /* new grammar rules */});
```
The problem is that because the baseGrammar's external rules are used in my new grammar. My new grammar requires its own scanner.c. 
How should I define scanner.c? 
I was thinking I just include the base grammar as a dev dependency, and then manually copy the contents of that base grammar's scanner.c to my new grammar. 
Alternatively, maybe I can create a .h file containing the inlined code use by the base grammar that both my new grammar and the base can consume the header.

I see tree-sitter-typescript had to make a decision on this and it seems like they went with copying javascript's scanner.c and modifying it from there. 

Thanks!

**Comment by ObserverOfTime** [comment]
See #4077
